File: Packages/GemRunCore/Sources/GameKitCore/ZoneRules.swift
import CoreModels
import Foundation

/// Zone selection + mint constants (docs/21) — the single source of truth
/// the selector, the progress tracker, tests, and (as documented values)
/// the future server pipeline all share, in the CollectionRules mold.
public enum ZoneRules {
    /// How many zones a day gets, when the map offers enough candidates.
    public static let maxZoneCount = 4
    /// How far out candidate anchors are considered.
    public static let searchRadiusM: Double = 2_500
    /// Zone circles are large on purpose — this is ground to cover, not a
    /// capture ring: 350–500 m radius vs the gems' 61 m.
    public static let minZoneRadiusM: Double = 350
    public static let maxZoneRadiusM: Double = 500
    /// A park must be at least this big (m²) to anchor a zone.
    public static let minParkAreaM2: Double = 8_000
    /// Polygon zones grow to at least this footprint (the min-radius
    /// circle's area): a small park's ring is scaled outward — odd shape,
    /// large area — capped at maxRingScale so a pocket park can't project
    /// a district.
    public static let targetZoneAreaM2: Double = .pi * 350 * 350
    public static let maxRingScale: Double = 3.0
    /// Vertex budget per stored zone ring (big traced parks carry
    /// hundreds; rendering and point-in-ring don't need them).
    public static let maxRingVertices = 160
    /// "Plenty of walkable trails": metres of strict pedestrian way that
    /// must run inside the circle when trail data is available.
    public static let minTrailLengthM: Double = 200
    /// Center-to-center distance between two picked zones must be at least
    /// this factor × the sum of their radii.
    public static let minSeparationFactor: Double = 1.2

    /// The mile: metres credited inside a zone that mint a card. (The app
    /// speaks miles everywhere; internals stay metric.)
    public static let mintDistanceM: Double = UnitFormat.metersPerMile
    /// GPS fixes worse than this never count (server anti-cheat rule,
    /// mirrored client-side).
    public static let maxAccuracyM: Double = 50
    /// Per-source speed caps: on the open map anything faster than a brisk
    /// walk is a vehicle; during a run, real runners hit 4–5 m/s.
    public static let mapSpeedCapMps: Double = 3.5
    public static let runSpeedCapMps: Double = 6.0
    /// Mirror of CollectionRules.teleportSpeed — sustained jumps reset the
    /// segment chain instead of crediting impossible metres.
    public static let teleportSpeedMps: Double = 8.0
    /// A gap longer than this between fixes breaks the segment chain
    /// (backgrounding, tunnel, app switch).
    public static let maxSampleGapS: TimeInterval = 120
    /// Segments shorter than this are GPS shimmer, not walking.
    public static let minSegmentM: Double = 3
    /// A zone stops minting after this many cards in one day.
    public static let maxMintsPerZonePerDay = 3
}

File: backend/api/zone_rules.py
"""Zone selection + mint constants — a 1:1 port of the iOS GameKitCore
ZoneRules (docs/21), the same way rules.py mirrors CollectionRules. The Swift
fixture tests double as vectors; if a value moves, move it on BOTH sides in
the same change.
"""
import math

# Selection
MAX_ZONE_COUNT = 4
SEARCH_RADIUS_M = 2_500.0
MIN_ZONE_RADIUS_M = 350.0
MAX_ZONE_RADIUS_M = 500.0
MIN_PARK_AREA_M2 = 8_000.0
# Polygon zones grow to at least the min-radius circle's footprint — odd
# shape, large area — capped so a pocket park can't project a district.
TARGET_ZONE_AREA_M2 = math.pi * 350 * 350
MAX_RING_SCALE = 3.0
MAX_RING_VERTICES = 160
MIN_TRAIL_LENGTH_M = 200.0
MIN_SEPARATION_FACTOR = 1.2

# The mile: metres credited inside a zone that mint a card (the app speaks
# miles everywhere; internals stay metric — UnitFormat.metersPerMile).
MINT_DISTANCE_M = 1_609.344
MAX_MINTS_PER_ZONE_PER_DAY = 3

# Per-segment gates (ZoneProgressTracker) — documented here for the future
# server-side track validation; the client enforces them today.
MAX_ACCURACY_M = 50.0
MAP_SPEED_CAP_MPS = 3.5
RUN_SPEED_CAP_MPS = 6.0
TELEPORT_SPEED_MPS = 8.0
MAX_SAMPLE_GAP_S = 120.0
MIN_SEGMENT_M = 3.0

# Draw salts (StableSeed) — "Zone" and "Mint" in ASCII, exactly the Swift
# literals, so independent draws never correlate and both sides agree.
ZONE_PICK_SALT = 0x5A6F6E65
MINT_SALT = 0x4D696E74
